numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -329,6 +336,107 @@ array_implement_array_function(
 }
 
 
+/*
+ * Implements the __array_function__ protocol for a Python function, as described in
+ * in NEP-18. See numpy.core.overrides for a full docstring.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_array_function(
+    PyObject *NPY_UNUSED(dummy), PyObject *positional_args)
+{
+    PyObject *implementation, *public_api, *relevant_args, *args, *kwargs;
+
+    if (!intern_strings()) {
+        if (PyErr_Occurred()) {
+            PyErr_SetString(PyExc_RuntimeError,
+                            "cannot load arrayfunction_override.");
+        }
+        return NULL;
+    }
+
+    if (!PyArg_UnpackTuple(
+            positional_args, "implement_array_function", 5, 5,
+            &implementation, &public_api, &relevant_args, &args, &kwargs)) {
+        return NULL;
+    }
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries.
+     */
+    if (PyDict_CheckExact(kwargs) &&
+            PyDict_Contains(kwargs, npy_arrayfunction_str_like)) {
+        PyDict_DelItem(kwargs, npy_arrayfunction_str_like);
+    }
+
+    return array_implement_array_function_internal(
+        implementation, public_api, relevant_args, args, kwargs);
+}
+
+
+/*
+ * Implements the __array_function__ protocol for a C function. Added as an extension
+ * to NEP-18 in an effort to bring NEP-35 to life with minimal dispatch overhead.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_c_array_function(
+    const char *function_name, PyObject *args, PyObject *kwargs,
+    PyObject **err_type, PyObject **err_value, PyObject **err_traceback)
```

Comment:
This is not a good way to handle this. The convention for functions like these is typically:
* Return something - success
* Return null with no exception set - no implementation found
* Return null with exception set - error

You can use `PyErr_Occurred()` at the caller to distinguish the last two cases.

Reply:
Thanks @eric-wieser , all three comments were addressed by https://github.com/numpy/numpy/pull/16935/commits/38e3e2254c7cc99c29e4b9559d103a9e362d968f .

Reply:
@eric-wieser I had missed this, right now I changed the logic to return NotImplemented (borrowed ref, but its private if we want to change that).  If you disagree with that change, please say so. Otherwise, I guess it will go in for now at least.